sympy/ntheory/ec.py

```diff
@@ -0,0 +1,396 @@
+from sympy.core.compatibility import as_int, is_sequence
+from sympy.core.numbers import oo
+from sympy.core.relational import Eq
+from sympy.core.symbol import symbols
+from sympy.polys.domains import FiniteField, QQ, RationalField, FF
+from sympy.solvers.solvers import solve
+from .factor_ import divisors
+from .residue_ntheory import polynomial_congruence
+
+
+
+class EllipticCurve:
+    """
+    Create the following Elliptic Curve over domain.
+
+    `y^{2} + a_{1} x y + a_{3} y = x^{3} + a_{2} x^{2} + a_{4} x + a_{6}`
+
+    The default domain is ``QQ``. If no coefficient ``a1``, ``a2``, ``a3``,
+    it create curve as following form.
+
+    `y^{2} = x^{3} + a_{4} x + a_{6}`
+
+    Examples
+    ========
+
+    References
+    ==========
+
+    [1] J. Silverman "A Friendly Introduction to Number Theory" Third Edition
+    [2] http://mathworld.wolfram.com/EllipticDiscriminant.html
+    [3] G. Hardy, E. Wright "An Introduction to the Theory of Numbers" Sixth Edition
+
+    """
+
+    def __init__(self, a4, a6, a1=0, a2=0, a3=0, modulus = 0):
+        if modulus == 0:
+            domain = QQ
+        else:
+            domain = FF(modulus)
+        self._domain = domain
+        domain = domain.convert
+        self.modulus = modulus
+        # Calculate discriminant
+        self._b2 = a1**2 + 4 * a2
+        self._b4 = 2 * a4 + a1 * a3
+        self._b6 = a3**2 + 4 * a6
+        self._b8 = a1**2 * a6 + 4 * a2 * a6 - a1 * a3 * a4 + a2 * a3**2 - a4**2
+        self._discrim = domain(-self._b2**2 * self._b8 - 8 * self._b4**3 - 27 * self._b6**2 + 9 * self._b2 * self._b4 * self._b6)
+        self._a1 = domain(a1)
```

Comment:
```suggestion
        self._a1 = a1
```
and similarly below.